### PR TITLE
middleware: Remove hard dependency on Gorilla Sessions

### PR DESCRIPTION
**Pull request by lstoll**
Right now using the middleware is hard bound to using gorilla's sessions library, which makes it annoying to use in any project that isn't already using that for sessions. This adds a more generic load/save implementation for a custom type, that consumers can implement with whatever their preferred session management scheme is.

A implementation of this Interface is provided for gorilla sessions, and it has been added in a way that is fully backwards compatible - existing code needs no changes. That has been marked as deprecated though, to indicate that we'll want to get rid of it at some point.

The existing tests exercise both the interface and the adapter.

### CODE REVIEW

File: middleware/middleware.go [hunk 1]
```diff
@@ -15,14 +16,29 @@ import (
...
-	sessionKeyOIDCState        = "oidc-state"
-	sessionKeyOIDCReturnTo     = "oidc-return-to"
-	sessionKeyOIDCIDToken      = "oidc-id-token"
-	sessionKeyOIDCRefreshToken = "oidc-refresh-token"
-)
+// Sessions are used for managing state across requests.
+type Sessions interface {
+	GetSession(*http.Request) (*SessionData, error)
+	SaveSession(http.ResponseWriter, *http.Request, *SessionData) error
+}
```
Comment (sr) [inline middleware/middleware.go:46]: nit, but WDYT of `type SessionStoreV2` with `Get` and `Put`? Less stutter-y. More generally I wonder if it's even worth maintaining backward compatibility in lstoll/oidc?

**Review by sr — APPROVED** [review]
rad

File: middleware/middleware.go (hunk 1 above)
Comment (lstoll) [reply middleware/middleware.go:46]: Would you change the value on the `Middleware` struct to `SessionStoreV2` as well?

Also I waffle a lot on the stutter thing on interfaces. Really generic stuff like `Get` often conflicts on implementations if you're passing something that does multiple things, I don't like having to have a sub-struct to work around that. But yeah, stutter.

Comment (sr) [reply middleware/middleware.go:46]: > Would you change the value on the Middleware struct to SessionStoreV2 as well?

Probably, yes.

> Really generic stuff like Get often conflicts on implementations 

I mean things like the Reader interface is also super generic, and isn't a problem in practice I dunno. https://pkg.go.dev/golang.org/x/crypto/acme/autocert#DirCache also.

Comment (lstoll) [reply middleware/middleware.go:46]: Fiiiiiine

Comment (lstoll) [reply middleware/middleware.go:46]: Dropped backwards compatibility, renamed things

File: middleware/memory_session.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/lstoll/oidc/internal"
+)
+
+// MemorySessionStore is a simple session store, that tracks state in memory. It
+// is mainly used for testing, it is not suitible for anything outside a single
```
Comment (sr) [inline middleware/memory_session.go]:
```suggestion
// is mainly used for testing, it is not suitable for anything outside a single
```

File: middleware/memory_session.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+	nc.Value = sid
+	m.sessions[sid] = *d
+
+	http.SetCookie(w, nc)
+
+	return nil
+}
+
+func (m *MemorySessionStore) init() error {
```
Comment (sr) [inline middleware/memory_session.go]: `NewMemoryStore(*http.Cookie) SessionStore` perhaps

Comment (lstoll) [reply middleware/memory_session.go]: Did that. Make the struct private too, not sure what having it public adds when it returns an interface.

**Pull request merged**